docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+requirements are then utilized by the cluster autoscaler to make decisions about the number
+of nodes to add during scale up operations.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated cloud providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and cloud providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Provide a mechanism for users to override the defaults for any given MachineSet or MachineDeployment.
+- Create an optional API contract that defines the minimum needed information for scaling from and to zero replicas.
+
+### Non-Goals/Future Work
+
+- Create an API contract that cloud providers must follow.
+- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
+- Create additional CRDs or controllers
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
+this information to create a prediction about how many nodes should be created when scaling. In
+most sitatuations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero sitatuation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the cloud provider.
+
+An optional field is proposed for users to add to their Infrastructure Machine Templates which will
+carry information about the CPU, memory, and GPU requirements for the machines created from the template.
+This field will then be copied to the status of any MachineSet or MachineDeployment with an infrastructure
+reference to the template during reconciliation. The cluster autoscaler will then read this information
+from the status field to perform its calculations.
+
+A user may override the field in the associated infrastructure template by applying annotations to the
+MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+By keeping this information with the MachineSet and MachineDeployment resource objects, we reduce the
+need for the cluster autoscaler to have a deep knowledge of Cluster API internals.
```

Comment:
> This is easier from an autoscaler point of view, but this could be potentially confusing from a user perspective, because the source of truth for those information is the InfrastructureMachine, while the override happens on a different resource.

i tend to agree, and i think if we decide as a group that the best solution here is for the autoscaler to read the infrastructure resource from the MachineSet|Deployment, that is a viable option as well. but, we need to ensure that we will keep the infrastructure ref that currently exists.

> Also, from a user perspective, what is the difference/the need for having annotation at two levels (MachineDeployment, MachineSet)? Are there precedence rules?

i've used the language of "MachineSet or MachineDeployment" because the autoscaler could be operating on either type, depending on how the user has configured the autoscaler. but in practice, the autoscaler won't be dealing with both types of objects for the same node group. the user would only need to annotate the object they have specified for autoscaling.

Reply:
i've been playing around with an implementation that does this copying and i'm starting to think it might be more appropriate for the autoscaler to extract the infrastructure template object if it has a reference, and then use annotations on the resource in question (eg MachineSet, MachineDeployment, and MachinePool somday) for an override mechanism.

so, from the capi side, a provider would only need to supply these fields on their infrastructure template and then the rest would work as expected. but, there would be no replication of the resource hint fields from the infrastructure ref to the status of the affected resource.